File: jwt-ruby/generate.rb
require 'jwt'
require 'optparse'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: generate.rb -k path/to/file -i your_issuer_id [-u a_usernamee]"
  opts.on("-h", "--help", "show help") do |h|
    options[:help] = opts.banner
  end

  opts.on("-k KEY", "--key KEY", "path to private key") do |key|
    options[:key] = key
  end

  opts.on("-i","--issuer_id ISSUER_ID","Your Issuer Id") do |issuer_id|
    options[:issuer_id] = issuer_id
  end
  opts.on("-u","--username USER_NAME", "A user name") do |user_name|
    options[:user_name] = user_name
  end
end.parse!

errors = []
[:key, :issuer_id].each do |key|
  unless options.has_key?(key)
    errors << "Option parameter %s is required" % key
  end
end

if errors.length > 0
  puts errors
  exit;
end

begin
  rsa_private_key = OpenSSL::PKey::RSA.new File.read options[:key]
rescue Errno::ENOENT => e
  puts "Caught the exception: #{e}"
  exit -1
end

current_time = Time.now.to_i

payload = {}
payload['iss'] = options[:issuer_id]
payload['iat'] = current_time
payload['exp'] = current_time + 1800

if options.key?(:user_name)
  payload['sub'] = options[:user_name]
end

token = JWT.encode payload, rsa_private_key, 'RS512'

puts token
# puts JWT.decode token, nil, false

File: jwt-python/generate.py
"""jwt_generate.

Usage:
  jwt_generate.py [--key=<str>] [--issuer=<str>] [--username=<str>]
  jwt_generate.py (-h | --help)

Options:
  --key=<str>       path to private key
  --issuer=<str>    issuer id from api connsole
  --username=<str>  username if generating user token
  -h --help         Show this screen.
"""

import jwt
import time
from docopt import docopt


def generate(arguments):

    del arguments['--help']

    proceed = True
    for k, v in arguments.items():
        if k == '--username':
            continue

        if v is None:
            print(f"parameter {k} is required")
            proceed = False

    if proceed is False:
        return 'sorry'

    privateKeyPath = arguments['--key']

    with open(privateKeyPath) as f:
        privateKey = f.read()

    currentTime = int(time.time())

    payload = {}
    payload['iat'] = currentTime
    payload['exp'] = currentTime + 1800
    payload['iss'] = arguments['--issuer']

    if '--username' in arguments:
        payload['sub'] = arguments['--username']

    # RS512 requires $ pip install cryptography
    encoded_jwt = jwt.encode(payload, privateKey, algorithm='RS512')
    return encoded_jwt


if __name__ == '__main__':
    arguments = docopt(__doc__)

    print(generate(arguments))
